src/EntityFrameworkCore.Generator.Core/Templates/MappingClassTemplate.cs

```diff
@@ -224,6 +224,30 @@ private void GenerateRelationshipMapping(Relationship relationship)
         }
         CodeBuilder.Append(")");
 
+        var primaryKeys = relationship.PrimaryProperties;
+        var nonPrimaryPrincipalKey = !primaryKeys
+            .Select(pp => relationship.PrimaryEntity.Properties.ByProperty(pp.PropertyName))
+            .All(p => p.IsPrimaryKey ?? true);
```

Comment:
Unsure what `IsPrimaryKey == null` would mean, so I used `?? true` so it would *not* generate the `.HasPrincipalKey` stuff in that case.  There may be other corner-cases I haven't considered as well.